### PR TITLE
feat: expose context identities

### PR DESCRIPTION
**Pull request by MatejVukosav**
Output example

GET {{url}}/admin-api/contexts/2GAJpURPC2cGYgSkZNjk3oLWwmca6CYyMf7j4CUcWLZA/identities
```
{
    "data": {
        "identities": [
            "4zvwRjXUKGfvwnParsHAS3HuSVzV5cA4McphgmoCtajS"
        ]
    }
}
```

### CODE REVIEW

File: crates/context/src/lib.rs [hunk 1]
```diff
@@ -238,6 +238,30 @@ impl ContextManager {
...
+
+        let mut iter = handle.iter::<calimero_store::key::ContextIdentity>()?;
+        let mut ids = Vec::<PublicKey>::new();
+
+        for (k, v) in iter.entries() {
+            let (k, v) = (k?, v?);
+
+            if k.context_id() == context_id {
+                if v.private_key.is_some() {
+                    ids.push(PublicKey::from(k.public_key()));
```
Comment (miraclx) [inline crates/context/src/lib.rs]: can we use this here:

```suggestion
                    ids.push(PublicKey(k.public_key()));
```

just so it breaks when I change the signature of `PublicKey` to be consistent with the pattern established with `ContextId`, `ApplicationId` and `BlobId`

the reason for this is `k.public_key()` "should" return `PublicKey` already, but currently doesn't

that'll change soon

File: crates/server/src/admin/handlers/context.rs [hunk 2]
```diff
@@ -13,8 +14,10 @@ use crate::admin::storage::client_keys::get_context_client_key;
 use crate::admin::utils::context::{create_context, join_context};
 
 #[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
 pub struct ContextObject {
     context: calimero_primitives::context::Context,
+    context_identities: Vec<String>,
```
Comment (miraclx) [inline crates/server/src/admin/handlers/context.rs]: will we always want all identities for the context when we're getting context metadata?

I think maybe this should be a different endpoint, wdyt?

File: crates/context/src/lib.rs [hunk 3]
```diff
@@ -238,6 +238,30 @@ impl ContextManager {
...
+        let handle = self.store.handle();
+
+        let mut iter = handle.iter::<calimero_store::key::ContextIdentity>()?;
+        let mut ids = Vec::<PublicKey>::new();
+
+        for (k, v) in iter.entries() {
+            let (k, v) = (k?, v?);
+
+            if k.context_id() == context_id {
+                if v.private_key.is_some() {
```
Comment (miraclx) [inline crates/context/src/lib.rs]: maybe introduce a `local_only` boolean parameter?

File: crates/server/src/admin/handlers/context.rs (hunk 2 above)
Comment (MatejVukosav) [reply crates/server/src/admin/handlers/context.rs]: I think it is relevant context metadata the same as app_id or last_tx_hash. 

Use case for this is to select with which identity you want to join selected context or in context details show context config data. 

Also, the payload should be small, as user won't have too much identities for a single context. I don't see such use case.

File: crates/context/src/lib.rs [hunk 4]
```diff
@@ -238,6 +238,45 @@ impl ContextManager {
...
+
+        for (k, v) in iter.entries() {
+            let (k, v) = (k?, v?);
+
+            if k.context_id() == context_id {
+                if !only_owned_identities || v.private_key.is_some() {
+                    ids.push(PublicKey(k.public_key()));
+                }
+            }
+        }
```
Comment (fbozic) [inline crates/context/src/lib.rs]: This iterates over all entries, but you already know part of the key. Why not use it, so we reduce in memory checks?
Reference: https://github.com/calimero-network/core/blob/1296f7d45974d63c3784d4d64fb29a38b5328652/crates/node/src/lib.rs#L566-L568

File: crates/server/src/admin/handlers/context.rs (hunk 2 above)
Comment (fbozic) [reply crates/server/src/admin/handlers/context.rs]: I agree this should be split. IMO we should avoid eagerly joining data in the server and allow clients to fetch it when needed.

File: crates/server/src/admin/handlers/context.rs [hunk 5]
```diff
@@ -1,5 +1,6 @@
 use std::str::FromStr;
 use std::sync::Arc;
+use std::vec;
```
Comment (miraclx) [inline crates/server/src/admin/handlers/context.rs]: is this unused?

File: crates/server/src/admin/handlers/context.rs [hunk 6]
```diff
@@ -34,12 +37,26 @@ pub async fn get_context_handler(
...
-            Some(context) => ApiResponse {
-                payload: GetContextResponse {
-                    data: ContextObject { context },
-                },
+            Some(context) => {
+                let context_identities = state
+                    .ctx_manager
+                    .get_context_owned_identities(context.id)
+                    .map_err(|err| parse_api_error(err).into_response())
+                    .unwrap_or_default()
```
Comment (miraclx) [inline crates/server/src/admin/handlers/context.rs]: looks to me like we're throwing this error away, yeah? we should have a `err.into_response` somewhere, unless you think it's irrelevant to share store-errors

File: crates/context/src/lib.rs (hunk 4 above)
Comment (MatejVukosav) [reply crates/context/src/lib.rs]: Thanks! Updated

File: crates/server/src/admin/handlers/context.rs (hunk 5 above)
Comment (MatejVukosav) [reply crates/server/src/admin/handlers/context.rs]: Doesn't mark as unused but also does not complain if I remove it.

there is below:
```
data: ContextUsers {
      context_users: vec![],
  },
```

File: crates/server/src/admin/handlers/context.rs (hunk 6 above)
Comment (MatejVukosav) [reply crates/server/src/admin/handlers/context.rs:45]: Idea is to not expose internal errors outside. So we map all errors to message we want to expose

Comment (MatejVukosav) [reply crates/server/src/admin/handlers/context.rs:45]: pub fn parse_api_error(err: eyre::Report) -> ApiError {
    match err.downcast::<ApiError>() {
        Ok(api_error) => api_error,
        Err(original_error) => ApiError {
            status_code: StatusCode::INTERNAL_SERVER_ERROR,
            message: original_error.to_string(),
        },
    }
}

Comment (MatejVukosav) [reply crates/server/src/admin/handlers/context.rs]: Do we really want to expose internal error outside? User can react only on invalid contextId. Not sure what else he can fix?

Comment (miraclx) [reply crates/server/src/admin/handlers/context.rs]: You're probably right, if this fails, for a database issue, then maybe `get_context` failed as well and we would've returned.

Just concerned about saying we don't have context IDs when we do, but had an internal error.

There's always a chance retrying might work, maybe..

File: crates/server/src/admin/handlers/context.rs (hunk 5 above)
Comment (miraclx) [reply crates/server/src/admin/handlers/context.rs]: Ah, yeah that's already in the standard library prelude. No need to import explicitly.

File: crates/context/src/lib.rs [hunk 7]
```diff
@@ -256,7 +256,21 @@ impl ContextManager {
...
+                break 'first None;
+            };
+
+            Some((k, iter.read()))
+        };
+
+        for (k, v) in first.into_iter().chain(iter.entries()) {
             let (k, v) = (k?, v?);
 
             if k.context_id() == context_id {
```
Comment (miraclx) [inline crates/context/src/lib.rs]: let's flip this. if it doesn't match, we can break

File: crates/server/src/admin/handlers/context.rs [hunk 8]
```diff
@@ -35,6 +33,44 @@ pub async fn get_context_handler(
         .get_context(&context_id)
         .map_err(|err| parse_api_error(err).into_response());
```
Comment (miraclx) [inline crates/server/src/admin/handlers/context.rs]:
```suggestion
        .map_err(|err| parse_api_error(err));
```

or

```suggestion
        ;
```
and ..._

File: crates/server/src/admin/handlers/context.rs [hunk 9]
```diff
@@ -35,6 +33,44 @@ pub async fn get_context_handler(
...
+                },
+            }
+            .into_response(),
+            None => ApiError {
+                status_code: StatusCode::NOT_FOUND,
+                message: "Context not found".into(),
+            }
+            .into_response(),
+        },
+        Err(err) => err.into_response(),
```
Comment (miraclx) [inline crates/server/src/admin/handlers/context.rs]: _...

```suggestion
        Err(err) => parse_api_error(err).into_response(),
```

**Review by miraclx — APPROVED** [review]
lgtm

**Review by frdomovic — APPROVED** [review]
Tested, all good!

File: crates/server/src/admin/handlers/context.rs (hunk 6 above)
Comment (MatejVukosav) [reply crates/server/src/admin/handlers/context.rs]: Make sense. Updated to return generic error if there is some internal issue

File: crates/context/src/lib.rs (hunk 7 above)
Comment (MatejVukosav) [reply crates/context/src/lib.rs]: Updated

**Review by miraclx — APPROVED** [review]

**Pull request merged**